shtab/__init__.py

```diff
@@ -275,85 +381,122 @@ def complete_bash(
-[ -n "$opts_more" ] && echo "$opts_more") )
+    current_action_nargs=1
   fi
+
+  current_action_args_start_index=$(( $word_index + 1 ))
+
+  current_action_is_positional=$2
 }
 
 # Notes:
-# `COMPREPLY` contains what will be rendered after completion is triggered
-# `word` refers to the current typed word
-# `${!var}` is to evaluate the content of `var`
-# and expand its content as a variable
-#       hello="world"
-#       x="hello"
-#       ${!x} ->  ${hello} ->  "world"
+# `COMPREPLY`: what will be rendered after completion is triggered
+# `completing_word`: currently typed word to generate completions for
+# `${!var}`: evaluates the content of `var` and expand its content as a variable
+#     hello="world"
+#     x="hello"
+#     ${!x} -> ${hello} -> "world"
 {root_prefix}() {
-  local word="${COMP_WORDS[COMP_CWORD]}"
-
+  local completing_word="${COMP_WORDS[COMP_CWORD]}"
   COMPREPLY=()
 
-  if [ "${COMP_CWORD}" -eq 1 ]; then
-    {root_prefix}_compgen_root_ ${COMP_WORDS[1]}
-  elif [ "${COMP_CWORD}" -eq 2 ]; then
-    {root_prefix}_compgen_command_ ${COMP_WORDS[1]}
-  elif [ "${COMP_CWORD}" -ge 3 ]; then
-    {root_prefix}_compgen_subcommand_ ${COMP_WORDS[1]} ${COMP_WORDS[2]}
+  prefix={root_prefix}
+  word_index=0
+  _set_parser_defaults
+  word_index=1
+
+  # determine what arguments are appropriate for the current state
+  # of the arg parser
+  while [ $word_index -ne $COMP_CWORD ]; do
+    local this_word="${COMP_WORDS[$word_index]}"
+
+    if [[ -n $subparsers && " ${subparsers[@]} " =~ " ${this_word} " ]]; then
+      # valid subcommand: add it to the prefix & reset the current action
+      prefix="${prefix}_$(_shtab_replace_nonword $this_word)"
+      _set_parser_defaults
+    fi
+
+    if [[ " ${current_option_strings[@]} " =~ " ${this_word} " ]]; then
+      # a new action should be acquired (due to recognised option string or
+      # no more input expected from current action);
+      # the next positional action can fill in here
+      _set_new_action $this_word false
+    fi
+
+    if [[ "$current_action_nargs" != "*" ]] && \\
+       [[ "$current_action_nargs" != "+" ]] && \\
+       [[ "$current_action_nargs" != *"..." ]] && \\
+       (( $word_index + 1 - $current_action_args_start_index >= \\
+          $current_action_nargs )); then
+      $current_action_is_positional && let "completed_positional_actions += 1"
+      _set_new_action "pos_${completed_positional_actions}" true
+    fi
+
+    let "word_index+=1"
+  done
+
+  # Generate the completions
+
+  if [[ "${completing_word}" == -* ]]; then
+    # optional argument started: use option strings
+    COMPREPLY=( $(compgen -W "${current_option_strings[*]}" -- "${completing_word}") )
+  else
+    # use choices & compgen
+    COMPREPLY=( $(compgen -W "${current_action_choices}" -- "${completing_word}"; \\
+                  [ -n "${current_action_compgen}" ] \\
+                  && "${current_action_compgen}" "${completing_word}") )
```

Comment:
not too sure about this...

```
[ -n "${current_action_compgen}" ] && "${current_action_compgen}" "${completing_word}"
```

so `current_action_compgen` is a command which takes one positional arg `completing_word`?

Reply:
Sorry about the late reply.  Your assessment is correct.  `current_action_compgen` will be a command such as `_shtab_compgen_files()` or `_shtab_compgen_dirs()` which takes one argument to generate possible completions.

Reply:
Is that okay?  Are there some custom compgen functions which aren't compatible with this setup?

Reply:
lgtm